lock/lock.go

```diff
@@ -0,0 +1,333 @@
+	"time"
+)
+
+// Terraform requires the DynamoDB table to have a primary key with this name
+const attributeLockId = "LockID"
+
+// Default is to retry for up to 5 minutes
+const maxRetriesWaitingForTableToBeActive = 30
+const sleepBetweenTableStatusChecks = 10 * time.Second
+
+const dynamoDbPayPerRequestBillingMode = "PAY_PER_REQUEST"
+
+type Options struct {
+	Logger *logrus.Logger
+	LockString string
+	LockTable string
+	AwsRegion string
+	MaxRetries int
+	SleepBetweenRetries time.Duration
+	CreateTableIfMissing bool
+}
+
+type TimeoutExceeded struct {
+	LockTable  string
+	LockString string
+	Timeout    time.Duration
+}
+
+func (err TimeoutExceeded) Error() string {
+	return fmt.Sprintf("Timeout trying to acquire lock %s in table %s (timeout was %s)\n", err.LockString, err.LockTable, err.Timeout)
+}
+
+type TableNotActiveError struct {
+	LockTable string
+}
+
+func (err TableNotActiveError) Error() string {
+	return fmt.Sprintf("Table %s is not active\n", err.LockTable)
+}
+
+// NewAuthenticatedSession gets an AWS Session, checking that the user has credentials properly configured in their environment.
+func NewAuthenticatedSession(awsRegion string) (*session.Session, error) {
+	sessionOptions := session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+		Config:            *aws.NewConfig().WithRegion(awsRegion),
+	}
+	sess, err := session.NewSessionWithOptions(sessionOptions)
+
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	if _, err = sess.Config.Credentials.Get(); err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	return sess, nil
+}
+
+// NewDynamoDb returns an authenticated client object for accessing DynamoDb
+func NewDynamoDb(awsRegion string) (*dynamodb.DynamoDB, error) {
+	sess, err := NewAuthenticatedSession(awsRegion)
+	if err != nil {
+		return nil, err
+	}
+	dynamodbSvc := dynamodb.New(sess)
+	return dynamodbSvc, nil
+}
+
+// AcquireLock will attempt to acquire a lock in DynamoDB table and will take the configuration options into account
+// We are using DynamoDB to crate a table to help us track the lock status of different resources.
+// The acquiring of a lock attempts to create a table if the configuration property `CreateTableIfMissing` is set to true,
+// otherwise it assumes that such a table already exists. The intention is that we have 1 table per resource in a single region.
+// This would allow the locking mechanism to flexibly decide if a resource is locked or not. For test cases where the AWS resource
+// is multi-region, or global, the configuration of which regions to use should reflect that.
+func AcquireLock(options *Options) error {
+	client, err := NewDynamoDb(options.AwsRegion)
+	if err != nil {
+		options.Logger.Errorf("Error authenticating to AWS: %s\n", err)
+		return err
+	}
+
+	if options.CreateTableIfMissing == true {
+		err := createLockTableIfNecessary(options, client)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
```

Comment:
NIT: A way to reduce the code for this pattern by one line:

```suggestion
		if err := createLockTableIfNecessary(options, client); err != nil {
			return errors.WithStackTrace(err)
		}
```